src/meta/processors/jobMan/MetaJobExecutor.cpp

```diff
@@ -145,42 +156,32 @@ ExecuteRet MetaJobExecutor::execute() {
     }
 
     std::vector<PartitionID> parts;
-    std::vector<folly::SemiFuture<Status>> futures;
+
     auto addresses = nebula::value(addressesRet);
+    auto unfinishedTask = addresses.size();
+
+    std::unordered_map<HostAddr, Status> ret;
     for (auto& address : addresses) {
         // transform to the admin host
-        auto future = executeInternal(Utils::getAdminAddrFromStoreAddr(address.first),
+        auto& host = address.first;
+        ret[host] = nebula::Status::Error();
+        auto fut = executeInternal(Utils::getAdminAddrFromStoreAddr(host),
                                       std::move(address.second));
-        futures.emplace_back(std::move(future));
-    }
-
-    std::vector<Status> results;
-    nebula::Status errorStatus;
-    folly::collectAll(std::move(futures))
-        .thenValue([&](const std::vector<folly::Try<Status>>& tries) {
-            for (const auto& t : tries) {
-                if (t.hasException()) {
-                    LOG(ERROR) << "Admin Failed: " << t.exception();
-                    results.emplace_back(nebula::Status::Error());
-                } else {
-                    results.emplace_back(t.value());
-                }
+        std::move(fut).thenTry([&](auto&& t) {
+            std::lock_guard<std::mutex> lk(muInterrupt_);
+            --unfinishedTask;
+            if (t.hasValue()) {
+                ret[host] = t.value();
             }
-        }).thenError([&](auto&& e) {
-            LOG(ERROR) << "Admin Failed: " << e.what();
-            errorStatus = Status::Error(e.what());
-        }).wait();
-
-    if (addresses.size() != results.size()) {
-        LOG(ERROR) << "hosts.size: " << addresses.size()
-                   << ", results.size: " << results.size();
-        return cpp2::ErrorCode::E_INVALID_PARM;
+            condInterrupt_.notify_one();
+        });
     }
 
-    std::unordered_map<HostAddr, Status> ret;
-    for (size_t i = 0; i != addresses.size(); ++i) {
-        ret.emplace(addresses[i].first, results[i]);
+    {
+        std::unique_lock<std::mutex> lk(muInterrupt_);
+        condInterrupt_.wait(lk, [&]{ return unfinishedTask == 0 || stopped_; });
```

Comment:
I wonder in what context this sync operation gets executed. Will it block the normal bussiness logics?

Reply:
job mgr run job one by one, and this where the current running job waiting for its sub task return.